ngraph/core/shape_inference/include/scatter_nd_base_shape_inference.hpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <openvino/op/util/scatter_nd_base.hpp>
+
+#include "utils.hpp"
+
+template <class T>
+void shape_infer(const ov::op::util::ScatterNDBase* op,
+                 const std::vector<T>& input_shapes,
+                 std::vector<T>& output_shapes,
+                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 3 && output_shapes.size() == 1);
+    const auto& inputs_shape = input_shapes[0];
+    const auto& indices_shape = input_shapes[1];
+    const auto& updates_shape = input_shapes[2];
+
+    const auto& inputs_rank = inputs_shape.rank();
+    const auto& indices_rank = indices_shape.rank();
+    const auto& updates_rank = updates_shape.rank();
+
+    NODE_VALIDATION_CHECK(op,
+                          indices_rank.is_dynamic() || indices_rank.get_length() >= 1,
+                          "Indices rank is expected to be at least 1");
+
+    NODE_VALIDATION_CHECK(op,
+                          inputs_rank.is_dynamic() || indices_rank.is_dynamic() ||
+                              indices_shape[indices_rank.get_length() - 1].is_dynamic() ||
+                              indices_shape[indices_rank.get_length() - 1].get_length() <= inputs_rank.get_length(),
+                          "Last dimension of indices can be at most the rank of inputs");
+
+    if (inputs_rank.is_static() && indices_rank.is_static() && updates_rank.is_static() &&
+        indices_shape[indices_rank.get_length() - 1].is_static()) {
+        auto expected_updates_rank = indices_rank.get_length() + inputs_rank.get_length() -
+                                     indices_shape[indices_rank.get_length() - 1].get_length() - 1;
+        // If expected updates rank is 0D it also can be a tensor with one element
+        NODE_VALIDATION_CHECK(op,
+                              updates_rank.get_length() == expected_updates_rank || expected_updates_rank == 0,
+                              "Rank of updates must be rank of inputs + rank of indices - last dimension of indices "
+                              "- 1");
+
+        bool compatible = true;
+        if (inputs_shape.is_static() && indices_shape.is_static() && updates_shape.is_static()) {
+            size_t static_indices_rank = indices_rank.get_length();
+            for (size_t i = 0; i < static_indices_rank - 1; i++) {
+                compatible = compatible && updates_shape[i].same_scheme(indices_shape[i]);
```

Comment:
this is a combo of requirement for shapes to be static, but using same_scheme to check the equality -- which is useful for dynamic shapes only. I propose to extend this code to work with dynamic shapes with static rank.